### ISSUE TITLE
MaskShader Bug

### ISSUE THREAD

**Issue by brannondorsey** [kibio/kibio#28]
![screen shot 2015-03-12 at 1 00 47 pm](https://cloud.githubusercontent.com/assets/3193533/6624571/2ab1d168-c8b8-11e4-89da-f92426bef97c.png)

When the Mask Tool is selected, and the mouse button is pressed a strange bug occurs where (see above). I've found that [this line](https://github.com/kibio/kibio/blob/25579e19d8e7fbb083aee9186f57687f3bacc47d/src/Layer.cpp#L200) in `Layer::draw()` is to blame:

```cpp
_maskShader.setUniformTexture("maskTex", _maskSurface.getTexture(), 1);
```

This seems to be a GLSL bug perhaps pertaining to the GL version which we are setting in `main.cpp`:

```cpp
#include "SimpleApp.h"

int main()
{
    ofGLWindowSettings windowSettings;
    windowSettings.setGLVersion(3,3);
    ofCreateWindow(windowSettings);

    ofRunApp(new Kibio::SimpleApp());
}
```

Our current GLSL shaders loaded into `_maskShader` are:

### mask.vert
```glsl
#version 150

// these come from the programmable pipeline
uniform mat4 modelViewProjectionMatrix;

in vec4 position;
in vec2 texcoord;

// texture coordinates are sent to fragment shader
out vec2 texCoordVarying;

void main()
{
    texCoordVarying = texcoord;
    gl_Position = modelViewProjectionMatrix * position;
}
```

### mask.frag
```glsl
#version 150

// these are our textures
uniform sampler2DRect tex0;
uniform sampler2DRect maskTex;

// this comes from the vertex shader
in vec2 texCoordVarying;

// this is the output of the fragment shader
out vec4 outputColor;

void main()
{
    // get rgb from tex0
    vec3 src = texture(tex0, texCoordVarying).rgb;

    // get alpha from mask
    float mask = texture(maskTex, texCoordVarying).r;
    
    //mix the rgb from tex0 with the alpha of the mask
    outputColor = vec4(src , mask);
}
```

These shaders seemed to work fine when we were using OF Master at or close to v0.8.4 and the method for setting the renderer was `ofSetCurrentRenderer(ofGLProgrammableRenderer::TYPE)`.

**Comment by bakercp** [comment]
But why only when the mouse is pressed?

**Comment by brannondorsey** [comment]
Not sure yet, still digging. Most likely it is linked directly to `Layer::_maskDirty` which triggers this code in `Layer::update()`:

```cpp
    if (_maskDirty)
    {
        _maskSurface.begin();
        ofClear(0, 0, 0, 0);
        ofPushStyle();
        ofSetColor(255);

        if (_mask && _mask->isAllocated())
        {
            _mask->draw(0, 0, _maskSurface.getWidth(), _maskSurface.getHeight());
        }
        else
        {
            ofDrawRectangle(0, 0, _maskSurface.getWidth(), _maskSurface.getHeight());
        }

        ofPopStyle();
        _maskSurface.end();

        _maskDirty = false;
  }
```

**Comment by brannondorsey** [comment]
However that just draws to the `_maskSurface` ofFbo so I can't think that would have anything to do with the strange translation applied by the `_maskShader` itself.

**Comment by brannondorsey** [comment]
Ah, here is the condition that causing the bug:

```
     if (ofGetMousePressed())
    {
        ofPoint layerMouse = screenToLayer(mouse);

        if (_parent.isMaskBrushEnabled())
        {
            _maskPath.clear();
            _maskSurface.begin();
            ofPushStyle();

            if (ofGetKeyPressed(OF_KEY_SHIFT))
            {
                ofSetColor(255);
            }
            else
            {
                ofSetColor(0);
            }

            _brushTex.draw(layerMouse.x - 25, layerMouse.y - 25, 50, 50);

            ofPopStyle();

            _maskSurface.end();
        }
    }
```

**Comment by brannondorsey** [comment]
Strangely however, drawing to it with `_brushTex.draw(layerMouse.x - 25, layerMouse.y - 25, 50, 50)` does not cause the bug (which rules out an error in the `screenToLayer(mouse)` coming from my transformation additions which I had previously thought might be causing this). It seems that simply the presence of `_maskSurface.begin()` and `_maskSurface.end()` cause this bug here.

**Comment by brannondorsey** [comment]
@bakercp, I've spent a while banging my head against the wall with this one. I think it might be beyond me. Everything keeps coming back to this line in `Layer::draw()`:

```
_maskShader.setUniformTexture("maskTex", _maskSurface.getTexture(), 1);
```

Everything works fine (except for the masking of course) without it, however, its such a basic shader and nothing looks wrong with it. Maybe I just don't know enough about what is going on under the hood.

**Comment by bakercp** [comment]
Yeah, it wasn't immediately obvious to me either.  I'd recommend using `git-bisect` to figure out exactly what commit broke it.  There may be something else going on that we're just not seeing ... if you can narrow it down to one commit, then I can jump in later tonight.

**Comment by brannondorsey** [comment]
Ok, using `git-bisect` I can confirm that the commit that broke this was a69e5fa261b5ecc5a15073259abf91e423224254. As I expected this was the commit where we changed from using:

```cpp
ofSetCurrentRenderer(ofGLProgrammableRenderer::TYPE)
```
to
```cpp
        ofGLWindowSettings windowSettings;
	windowSettings.setGLVersion(3,3);
	ofCreateWindow(windowSettings);
```

The tricky part is, that we've been compiling against an in-flux master branch, and this commit also requires a large jump in the OF Master branch we were compiling against. I tested everything up until the broken commit using OF Master at 3b6bf40a097de5bca0e2891b6bf3e6e2c59a51c3 which is a commit from Dec 1, 2014. I picked this OF Master commit because it is about even with our first release and workshop. Our broken Kibio commit will no longer work with that OF Maser commit, so instead I am using OF Master at 17e576e0c9977d3d1d75a170b8fad1ecc72076ab (which is a commit from March 6). I didn't want to use `develop-apothecary` because I am not familiar with its contents and didn't want to introduce a new variable.

I'm not sure at which OF commit `ofSetCurrentRenderer(ofGLProgrammableRenderer::TYPE)` was removed, but it cannot be used with our broken Kibio commit. To reproduce this test you will need to use CMD-Click on the layers in order to activate the mask tool. I suggest checking out the commit right before our broken one (d26a33ff7a0c847caa4f41e5cd26ef9108dba2e2) to see what the behavior is supposed to be like. I hope this is enough information.
